### ISSUE TITLE
FEATURE: Would like to add Few more controllers or API's  to the Chat-app

### ISSUE THREAD

**Issue by arnb-smnta** [hiteshchoudhary/apihub#102]
### Describe the Feature

I am proposing to add  more controllers or API function in the chat-app

I want to add
1.Delete a single message or attachment in one to one chat within 15 mins of creation of the message
2.Upload group Picture  by the admins
3.Upload description of groups by the admins
4.Editting the group picture or group description by the admins
5.Deleting messages in a group by the sender within 15 minutes of sending
6.Creating different types of group where only admins can send the messages and attachments  and participants can see

### Use Case

This will enhance the chat app experience for the users and also improve the functionality of the chat app like in the professional apps like whatsapp.

### Additional Information

@hiteshchoudhary 
I would like to start working on the additional api's in the chat app i was trying to make my own chat-app with free api but found those functionalities missing so i would like to add them .

### Suggested Tools

Technologies
No additional technologies other than those that are already being used

**Comment by wajeshubham** [comment]
@arnb-smnta look interesting, let's discuss more on the features. Can you elaborate more on each feature and timeline that we are looking for? Thanks

**Comment by arnb-smnta** [comment]
@wajeshubham  
1.[Delete/chat-app/chats/{chatid}/{messageid}]I will like to add delete message route/controller for the chat app-that will delete a particular message sent by the user in one to one chat within 15 min of creation of the message.

2.[Delete/chat-app/chats/group/{chatid}/{messageid}]This route will delete message in the group chat the message can be deleted by the user who sent the message within 15 mins of creation or by the admin of the group at any time .

3. Modify-[Post/chat-app/chats/group]-It will also take a single group profile pic (optional) during creation of the group along with groupname(already coded) and description in the body(will add) and participants(already coded) and type of group(will add) .(Group profile pic will be stored in CLoudinary)

Necessary changes  will add for this route are
   1.Upload on cloudinary function 
   2. Delete on cloudinary function 
   3. required env files for cloudinary 
   4. Modification of chat model will add a profile pic:{type:{string || ""},description:{type:String}, groupType: { type: String, enum:["adminOnly","everyone"],default: null, // Default to null for one-on-one chats}

4.[Patch/chat-app/chats/group/pic/{chatid}]-This route will be used to change /add Profile Pic of the group (Admins will only have the rights to change group pro pic)
5.[Delete/chat-app/chats/group/pic/{chatid}]-This route will be used to delete profile pic of the group(only admins can change it)

6.Modify[Patch/chat-app/chats/group/{chatid}]-It takes new group chat name for updation ,will update it to take group chat description along with chat name  and type of group for updation.(only admins will have the right)

7.[Patch/chat-app/chats/group/admin/{chatid}/{participantid}]-To add the participants as admin . (can be done by any other admin)

Will have to modify the chat model from    admin: {
      type: Schema.Types.ObjectId,
      ref: "User",
    },  to
admin:[{
      type: Schema.Types.ObjectId,
      ref: "User",
    },]  to accomodate array of admins.

8.[Delete/chat-app/chats/group/admin/{chatid}/{participantid}]-To remove Someone as group's admin -(can be done only by the first member or the creator of the group if the creator removes himself as admin then there must be a second admin(required) to be the primary admin or it will throw an error)


@wajeshubham can i also modify user controller because the model has avatar pic built in it but the controller code for updating user avatar or accepting user avatar while registering is not written (if i update it it might affect all the other APIs because User is used by eveyone) but with my cloudinary functions i can store the avatar pics in cloudinary and also delete them while updation If you allow I can update user registration route/controller for taking user avatar i will keep it optional so that it does not effect the existing API's and will also add the modify avatar route/controller.

**Comment by arnb-smnta** [comment]
hey subham i have submitted my reply can u please go through it

**Comment by arnb-smnta** [comment]
@wajeshubham 
Sir kindly please review the features I have proposed and can i start working on the features or do i need to provide more clarification on the routes,

**Comment by wajeshubham** [comment]
@arnb-smnta Thanks for the interests but please consider following points:

- We are avoiding usage of any cloud object or file storage as of now because we want to keep the file storage consistent throughout the application. Right now our files are stored on the disk inside the public folder which we are planing to more to the cloud but that's a future plan. So, we cannot allow integration of cloudinary as of now.

- Regarding admin functionality, admin can delete message/attachment of the use is a bit redundant as we are not building any specific chat app but a generic one where we don't want to complicate the 1st version by introducing role based access for certain functionality but this is a good feature for the future versions. 

- I don't think idea of admin only groups also fits the current context of the FreeAPI project. We want to keep apis as self explanatory and lean as possible but as I said, I like the idea of role based access logic so we will definitely consider this feature if we plan to add more complex functionality in the future versions.

- Most importantly we don't recommend updating user model to introduce any functionality in other modules as the user model is common and shared between modules.

In conclusion, the feature that we liked and would like you to contribute is the delete message feature after 15 minutes by the authenticated user who is the owner of that message only. So, you can proceed with implementing that endpoint in your forked repository's branch. In case of any query, please feel free to ask.

**Comment by arnb-smnta** [comment]
Hi @wajeshubham 
1.I will go ahead with the following api end point then [Delete/chat-app/messages/{chatid}/{messageid}]- delete message route/controller for the chat app-that will delete a particular message or attachement sent by the user in one to one chat and group chat  within 15 min of creation of the message only by the sender . 

You can assign this  task to me.

Question-Should i fire a Message_Deleted_Event for every partcipant along with user to update the chat after deletion.

A query if i keep the assets in server can i go ahead and also 

1.Modify-[Post/chat-app/chats/group]-It will also take a single group profile pic(stored in the server) (optional) during creation of the group along with groupname(already coded) and description in the body(will add) and participants(already coded) .

2.[Patch/chat-app/chats/group/pic/{chatid}]-This route will be used to change remove /add Profile Pic of the group (Admins will only have the rights to change group pro pic)

3.Modify[Patch/chat-app/chats/group/{chatid}]-It takes new group chat name for updation ,will update it to take group chat description along with chat name for updation.(only admins will have the right)


I can also code the frontend after doing the backend part to accept profile pic and description in the chat-app 

Conclusion: I am going ahead with creating the deletion API as suggested by you , do let me know about the other suggestions i have given and in the future versions if you ever consider adding other functionalities i have already made those functionalities for my own chat app -backend will be happy to contribute to those features as well .

**Comment by arnb-smnta** [comment]
@wajeshubham any updates

**Comment by wajeshubham** [comment]
PR: #109
